### ISSUE TITLE
Improve Makefile --dry-run output parsing

### ISSUE THREAD

**Issue by AntonioL** [microsoft/vscode-makefile-tools#185]
I am dealing with legacy code which issues compile commands in the form of:
```
ENV_VAR_1=XX ENV_VAR_2=YY g++ ....
```

This makes vscode-makefile-tools not parse the include paths, somehow, so I guess it fails in parsing the dry-run output. If you tell me where are the logs I can check that and provide that for you. 

I propose that --dry-run output parsing takes care of eventual env vars being prepended to the command.

I patched our Makefiles manually to avoid such environment variables, which are not needed, by being prendend but it would be best if vscode-makefile-tools would do so!

**Comment by AntonioL** [comment]
I can confirm this is the case by looking at the configurationCache

**Comment by andreeis** [comment]
@AntonioL, thank you for this feature idea that we overlooked.
Since I didn't work much with this scenario and you have more experience, please help me understand in more detail.
- In the case you wrote above, ENV_VAR_1 and ENV_VAR_2 would apply only to the g++ invocation of that line and not anytime below, right?
-can those ENV_VAR_1 and ENV_VAR_2 be used in the g++ arguments (for -D, -I...etc compiler swithces?) and the extension should expand them?

To unblock you with a workaround (other than rewriting your makefiles), do you have a wild set of different variables and combinations being set? I am thinking that you can produce manually in a terminal a full clean build log (redirect output of a full clean make build), then replace "ENV_VAR_1=XX ENV_VAR_2=YY" (and any similar strings) to empty string "" and let the extension read from that build log via "makefile.buildLog" in your .vscode/settings.json. Just make sure that when you add/remove source files or add/remove/modify compilation switches that may be relevant for IntelliSense, you just regenerate the build log to operate on latest information.

**Comment by AntonioL** [comment]
> 
> 
> @AntonioL, thank you for this feature idea that we overlooked.
> Since I didn't work much with this scenario and you have more experience, please help me understand in more detail.
> 
>     * In the case you wrote above, ENV_VAR_1 and ENV_VAR_2 would apply only to the g++ invocation of that line and not anytime below, right?
> 
>     * can those ENV_VAR_1 and ENV_VAR_2 be used in the g++ arguments (for -D, -I...etc compiler swithces?) and the extension should expand them?
> 
> 
> To unblock you with a workaround (other than rewriting your makefiles), do you have a wild set of different variables and combinations being set? I am thinking that you can produce manually in a terminal a full clean build log (redirect output of a full clean make build), then replace "ENV_VAR_1=XX ENV_VAR_2=YY" (and any similar strings) to empty string "" and let the extension read from that build log via "makefile.buildLog" in your .vscode/settings.json. Just make sure that when you add/remove source files or add/remove/modify compilation switches that may be relevant for IntelliSense, you just regenerate the build log to operate on latest information.

That is correct in the above example the environment variables would apply only to that g++ invocation on that line.
I did not know about that makefile.buildLog param, I can investigate that, will do that tomorrow and let you know.

Taking aside the workarounds we are discussing.
What do you think @andreeis should this extension be in the business of figuring out the env vars or not?

**Comment by andreeis** [comment]
> Taking aside the workarounds we are discussing.
> What do you think @andreeis should this extension be in the business of figuring out the env vars or not?
> 

Ideally yes, if we find out that they can be used in the compiler command. The extension is not invoking the compiler and since those environment variables don't matter below, this is an argument for not calculating them, only to ignore them so that we can parse correctly the rest of the line. But if the compiler uses switches by expanding those variables set before, then we would need to calculate them for the expansion.
We can triage that based on how often we hear the request from the community. Anyone can leave a thumbs up on the first entry of the GitHub issue.

The extension already does some environment variables special handling when running the "makefile.preconfigure" command by persisting in the VSCode instance environment anything impacted by running the "makefile.preconfigureScript" file. But this is different then the temporary on-the-fly env vars defined before a compiler invocation.

Also, we would later support other scenarios of variable substitution but that would be for example in settings.json. Would be different than substituting a temporary variable in a compilation command.

Just noting all the similar concepts that we address or plan to, for clarity.

**Comment by AntonioL** [comment]
I can confirm that indeed patching the "makefile.buildLog" works like a charm. It is not great dev experience but yeah it definitely works!

Looking forward to this feature being prioritised!

**Comment by andreeis** [comment]
@AntonioL, if you notice anything that is not working (other than when modifying source files/switches in the makefile which require build.log regeneration) let us know because besides not being able to refresh itself everything else should still work: configure, IntelliSense, having build targets and launch targets identified in the left side panel, building, debugging, running in terminal... all the "makefile." commands...

**Comment by AntonioL** [comment]
@andreeis I will review and update here.

**Comment by AntonioL** [comment]
> 
> 
> @AntonioL, if you notice anything that is not working (other than when modifying source files/switches in the makefile which require build.log regeneration) let us know because besides not being able to refresh itself everything else should still work: configure, IntelliSense, having build targets and launch targets identified in the left side panel, building, debugging, running in terminal... all the "makefile." commands...

I got a chance to play around with this for a while and so far I cannot note regressions when leveraging makefile.buildLog .
